### ISSUE TITLE
CAPI module which creates token gates access

### ISSUE THREAD

**Issue by jainkrati** [paritytech/capi#222]
shouldGetAccess(List<TokenConditions>) returns True/False for granting access to a resource on / off chain depending on whether the conditions are met. Ex: Grant access if user hold 100 tokens of Type A and min 20 tokens of Type B.

**Comment by harrysolovay** [comment]
I like the idea of standardizing a flow for validation rule definition/composition. A few high-level questions:

- Should we tackle this level of specificity (token thresholds)?
- What patterns are common in existing projects?
  > **Might** be useful to skim [the Subsocial parachain code](https://github.com/dappforce/subsocial-parachain).
- Where do you see this being most useful? Browser app code could not rely on this to guard anything mission critical... JS services might stand to benefit.
- Does this seem like a should-be member of the Capi std lib, or a should-be community lib (kind of like [GraphQL Shield](https://www.graphql-shield.com/))? I'd imagine the community will build out all kinds of validations.

Looking at `shouldAllowAccess`, it feels to me like a more fitting name might be `accountBalanceGte`. Is this correct?

```ts
import * as C from "capi";
import * as G from "./generated.ts";

function accountBalanceGte<
  PublicKey extends C.$<Uint8Array>,
  Threshold extends C.$<number>,
>(
  publicKey: PublicKey,
  threshold: Threshold,
) {
  return C.atom([
    G.system.account.get(addr),
    threshold,
  ], ({ data: { free } }, threshold) => {
    return a.data.free >= threshold;
  });
}
```

... following this train of thought:

- how might such a predicate vary between chains?
- might devs want to adjust the logic (gt, gte, fetch-some-data-to-compare, etc.)?

My sense is that users will define their own predicates. That being said, I can imagine this being extremely useful in cases where the threshold value is dynamic and stored on chain.

```ts
// ... continued ...

function accountBalanceGteThreshold<PublicKey extends C.$<Uint8Array>>(
  publicKey: PublicKey,
) {
  return accountBalanceGte(
    PUBLIC_KEY,
    G.somePallet.thresholds.get(PUBLIC_KEY),
  );
}
```

What other similar use cases might be desirable?

Pinging @kratico, as I'm curious to hear his thoughts as well :)

**Comment by jainkrati** [comment]
I agree that there is more flexibility for developers if we allow them to define their own conditions for token gating. 

I also like the idea you proposed earlier about capturing threshold values stored on chain and supporting blocklists for token gating features. I will add the same in the first post.

**Comment by kratico** [comment]
From my understanding, token gating could unlock features at different levels
- in the blockchain, to unlock/allow a blockchain feature or an extrinsic (this is similar to GQL Shield server-side permissions)
- in the client, to unlock/allow a UI/client feature

Capi could help in the UI/client side with composable utils in the std-lib or community-lib.

I'm not sure what would be the best source to identify utils like `accountBalance, accountBalanceGte, etc.`

**Comment by harrysolovay** [comment]
Closing in favor of #596
